libs/libcommon/src/libcommon/orchestrator.py

```diff
@@ -577,10 +615,58 @@ def _create_plan(self) -> None:
             self.add_task(CreateJobsTask(job_infos=job_infos_to_create))
 
 
+@dataclass
+class DatasetRemovalPlan(Plan):
+    """
+    Plan to remove a dataset.
+
+    The plan is composed of tasks to delete jobs and cache entries.
+
+    Args:
+        dataset: dataset name
+    """
+
+    dataset: str
+
+    def __post_init__(self) -> None:
+        super().__post_init__()
+        self.add_task(DeleteDatasetJobsTask(dataset=self.dataset))
+        self.add_task(DeleteDatasetCacheEntriesTask(dataset=self.dataset))
+
+
 @dataclass
 class DatasetOrchestrator:
+    """
+    Orchestrator for a dataset.
+
+    Args:
+        dataset (str): The name of the dataset.
+        processing_graph (ProcessingGraph): The processing graph.
+        blocked_datasets (list[str]): The list of blocked datasets. Supports Unix shell-style wildcards.
+
+    Raises:
+        libcommon.exceptions.DatasetInBlockListError: If the dataset is in the block list. As a side-effect, the
+          dataset is deleted from the Datasets Server.
+    """
+
     dataset: str
     processing_graph: ProcessingGraph
+    blocked_datasets: InitVar[list[str]]
+
+    def __post_init__(self, blocked_datasets: list[str]) -> None:
+        try:
+            raise_if_blocked(dataset=self.dataset, blocked_datasets=blocked_datasets)
+        except DatasetInBlockListError:
+            logging.warning(f"The dataset {self.dataset} is in the block list, we delete it from the Datasets Server.")
+            self.remove_dataset()
+            raise
```

Comment:
Maybe only raise when calling backfill or finish_job or set_revision or any other function that would create a job ? Otherwise you can't create one to call `remove_dataset` on a blocked dataset

Anyway in general a `__post_init__` raise is not always convenient e.g. for testing or if you need to call some read methods even on a blocked dataset for some reason.

Reply:
> Otherwise you can't create one to call remove_dataset on a blocked dataset

I expose `remove_dataset` for other admin endpoints (but I don't use it now, so...)

> Anyway in general a __post_init__ raise is not always convenient

yes, you're right, it's a bad practise to raise in the constructor (or like here, in __post_init__). I'll change this.

Reply:
https://github.com/huggingface/datasets-server/pull/1899/commits/125789acd3fa8db89b699f01d20465011bfc848f